internal/cloud/realm/client.go

```diff
@@ -93,54 +93,27 @@ func (c *client) do(method, path string, options api.RequestOptions) (*http.Resp
 	client := &http.Client{}
 	res, err := client.Do(req)
 	if err != nil {
-		return res, err
+		return nil, err
 	}
-	if res.StatusCode != http.StatusUnauthorized || !options.UseAuth || options.PreventRefresh {
-		return res, err
+	if res.StatusCode >= 200 && res.StatusCode <= 299 {
+		return res, nil
 	}
 
-	serverError := unmarshalServerError(res).(ServerError)
-	if serverError.Code != InvalidSessionCode {
-		return res, err
+	serverError, ok := unmarshalServerError(res).(ServerError)
```

Comment:
we shouldn't be returning `res, nil` down below at all.  once we're at this point in the code, we know we don't have a successful response so _some_ error needs to be returned.  a couple of changes here:

- it probably makes sense to change the existing name: `unmarshalServerError` to something like `parseResponseError`
- **IF** the error is not of type `ServerError` **OR** its error code is not an `invalidSessionCode`, then simply return `nil, parsedErr`
- **ELSE** do the retry logic, and if an error comes from there we should return an invalid session error

something like:

```go
parsedErr := parseResponseError(res)
if err, ok := parsedErr.(ServerError); !ok {
  return nil, parsedErr
} else if err.Code != invalidSessionCode {
  return nil, e
}
session, sessionErr := c.refreshAuth()
if sessionErr != nil {
  return nil, ErrInvalidSession
}
c.session = session
options.PreventRefresh = true
return c.do(method, path, options)
```

Reply:
Ah yeah this is a really good point. Will make that change